src/containers/Dashboard.js

```diff
@@ -0,0 +1,163 @@
+    this.fetchStoredData = this.fetchStoredData.bind(this);
+  }
+
+  componentDidMount() {
+    initData(this.state.chartData.datasets[0].data);
+  }
+
+  render() {
+    const showError =
+      this.state.display === Display.ERROR ? 'visible' : 'invisible';
+    const showLoading =
+      this.state.display === Display.LOADING ? 'visible' : 'invisible';
+    const showDashboard =
+      this.state.display === Display.LOADING ||
+      this.state.display === Display.ERROR
+        ? 'none'
+        : 'block';
+    return (
+      <Container>
+        <Row className='justify-content-center mt-5'>
+          <h1>Dashboard App</h1>
+        </Row>
+        <Row className='justify-content-center mb-5 mt-5'>
+          <Button
+            variant='info'
+            size='sm'
+            onClick={this.generateLocalData}
+            style={{ margin: 10 }}
+            active={this.state.display === Display.FRONTEND}
+          >
+            Generate data in frontend
+          </Button>
+
+          <Button
+            variant='info'
+            size='sm'
+            onClick={this.getRemoteData}
+            style={{ margin: 10 }}
+            active={this.state.display === Display.BACKEND}
+          >
+            Generate data in backend
+          </Button>
+
+          <Button
+            variant='info'
+            size='sm'
+            onClick={this.fetchStoredData}
+            style={{ margin: 10 }}
+            active={this.state.display === Display.DB}
+          >
+            Fetch data from database
+          </Button>
+        </Row>
+        <Container style={{ display: showDashboard }}>
+          <Row style={{ marginBottom: 100 }}>
+            <Col className='col-4'>
+              <Chart type='doughnut' data={this.state.chartData} />
+            </Col>
+            <Col className='col-4'>
+              <Chart type='pie' data={this.state.chartData} />
+            </Col>
+            <Col className='col-4'>
+              <Chart type='bar' data={this.state.chartData} />
+            </Col>
+          </Row>
+          <Row>
+            <Col className='col-4'>
+              <Chart type='line' data={this.state.chartData} />
+            </Col>
+            <Col className='col-4'>
+              <Chart type='polarArea' data={this.state.chartData} />
+            </Col>
+            <Col className='col-4'>
+              <Chart type='horizontalBar' data={this.state.chartData} />
+            </Col>
+          </Row>
+        </Container>
+        <Container className={showError}>
+          <Alert variant='danger'>Oh snap! You got an error!</Alert>
+        </Container>
+        <Container fluid className={showLoading}>
+          <Spinner animation='border' variant='info' className='mr-2' />
+          <span className='pl-1'>Loading...</span>
+        </Container>
+      </Container>
+    );
+  }
+
+  /* Generate random chart data */
+  generateLocalData() {
+    const newdata = Object.assign({}, this.state.chartData);
```

Comment:
Again, this doesn't perform a deep copy of the object.
Try this in your browser console:
```js
const chartData = { datasets: [ { data: [1,2,3] } ] };
const newdata = Object.assign({}, chartData);
newdata.datasets[0].data = [3,4,5];
chartData.datasets[0].data;
```

You'll see that the original chartData was modified.

I suggest one of these versions:
```js
import produce from 'immer';

const newData = produce(this.state.chartData, draftState => {
  draftState.datasets[0].data = [/* ... */];
})
```
Or
```js
const newData = {
  ...this.state.chartData,
  datasets: [
    {
      ...this.state.chartData.datasets[0],
      data: [/* ... */],
    },
  ],
};
```
Which is the equivalent of
```js
const newData = Object.assign({}, this.state.chartData, {
  datasets: [
    Object.assign({}, this.state.chartData.datasets[0], { data: [/* ... */] }),
  ],
};
```

Reply:
Sorry, I missed this comment in your prev review. 
Thanks for the examples! I decided to keep it simple and not use additional npm package. 
I fixed by using` ...this.state.chartData` .